ADR square size independent message commitments (#835)
celestiaorg/celestia-app: docs/architecture/adr-008-blocksize-independent-commitment.md

# ADR 008: Blocksize independent Commitments for Messages

## Changelog

- 03.08.2022: Initial Draft

## Context

> Currently, commitments in Celestia are dependent on the size of the block. The following diagram visualizes how this can happen. The visualizations in this ADR are without parity shares/erasure encoding.

Reviewer: just to note that removing the neccessity to malleate transactions depending on square size is a massive amount of work, considering that this is a major responsibility of the application. That's **not** to say that we shouldn't!! this is purely a comment

Author: Should i add this as a neutral consequence, because it will take a lot of work but could simplify the design. ?

> Currently, commitments in Celestia are dependent on the size of the block. The following diagram visualizes how this can happen. The visualizations in this ADR are without parity shares/erasure encoding.

Reviewer 2: ```suggestion
Currently, commitments in Celestia are dependent on the size of the square. The following diagram visualizes how this can happen. The visualizations in this ADR are without parity shares/erasure encoding.
```

The yellow block in the diagram is a message of length 11 shares. A square of 4x4 results in the following subtree roots: `A B C D`.
In [`CreateCommit`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L112-166), we calculate the commitment of the subtree roots with [`HashFromByteSlices`](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/crypto/merkle/tree.go#L7-L21). This resulting merkleroot is `Com`.
A square of 8x8 results in the following subtree roots: `H1 C D`
[`HashFromByteSlices`](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/crypto/merkle/tree.go#L7-L21) results now in the merkleroot `Com'`, which is not equal to `Com`.

![Old Commitment](./assets/size-dependent-commitment.png)

> To have a commitment independent of the block size, you would need to create a merkle tree over subtree roots that are included in every possible block size.

Reviewer 2: ```suggestion
To have a commitment independent of the square size, you would need to create a Merkle tree over subtree roots that are included in every possible square size.
```

> We agree on producing the commitment as if the message was put into the `minimalBlocksize`. Therefore the subtree roots will be the same and the resulting commitment as well, no matter the block size. *This is because all the subtree roots in the `minimalBlocksize` are also included in bigger block sizes*.

Reviewer 2: ```suggestion
We agree on producing the commitment as if the message was put into the `minimumSquareSize`. Therefore the subtree roots will be the same and the resulting commitment as well, no matter the square size. *This is because all the subtree roots in the `minimumSquareSize` are also included in bigger square sizes*.
```

Reviewer 2: [optional] it may help future readers if this section states that the diagram below assumes `minimumSquareSize=4`

You can see this in the updated version of the diagram above. Now the merkle root will stay the same. This property results from the block width and length being a power of 2.  

![New Commitment](./assets/size-independent-commitment.png)

You can see in the diagramm below that no matter how the message is split up into rows, the hashes of the subtree roots stay the same.

![Row Size Comparison](./assets/row-size-comparison.png)

> This scheme works for interactive commitments as well as long as the index of when a message starts is the same in every block size. In other words, the leaf nodes and leaf node pairs are independent of the block size.

Reviewer 2: [nit]

```suggestion
This scheme also works for interactive commitments as long as the index of when a message starts is the same in every block size. In other words, the leaf nodes and leaf node pairs are independent of the square size.
```

For example, in the following diagram, you see the message starting at the second share and still having the same commitment. I marked the skipped blocks pink to show that both messages have the same starting index.

![Interactive Commitment](./assets/interactive-commitment.png)

> If the message starting point index is larger than the row of `minimalBlocksize`, you take the index mod `minimalBlocksize`. The `minimalBlocksize` is shown in green, and the skipped block is in pink. In both cases, the commitment stays the same.

Reviewer 2: ```suggestion
If the message starting point index is larger than the row of `minimumSquareSize`, you take the index mod `minimumSquareSize`. The `minimumSquareSize` is shown in green, and the skipped block is in pink. In both cases, the commitment stays the same.
```

![Interactive Commitment 2](./assets/interactive-commitment2.png)

> 

Author: I was thinking of using old non-interactive default rules. I did not realize we could actually save padding using the index that is a multiple of the minimumSquareSize.

## Alternative Approaches

> TODO

Reviewer: can we link ADR007 from rollmint here?

Reviewer 2: [optional] to resolve @evan-forbes 's comment

```suggestion
## Alternative Approaches

- [rollmint/adr-007](https://github.com/celestiaorg/rollmint/blob/cb5c7440a8e879778e71097e254c3dd692c39d14/docs/lazy-adr/adr-007-header-commit-to-shares.md)
```

## Decision

TODO

> ## Detailed Design

Reviewer 2: [no change needed] we don't need to tackle in this PR but wanted to note this: we may consider including the `minimumSquareSize` in the block header so that we can increase this value in the future (potentially via governance proposal). 

If we don't include it in the block header, it would be difficult (perhaps impossible) to determine the `minimumSquareSize` for historical blocks.

Reviewer 2: This is resolvable since we'll use a minimumSquareSize that is message dependent. Perhaps `msgMinimumSquareSize` is a more appropriate name to disambiguate from `appconsts.MinSquareSize`

To implement this decision, you need to change [`CreateCommit`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L112-166).
In Detail, [`powerOf2MountainRange`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L142) should take `minimalBlocksize` as an argument instead of `squareSize`.

> Following the non-interactive default rules, `minimalBlocksize` can be calculated like this:

Reviewer 2: ```suggestion
Following the non-interactive default rules, `minimumSquareSize` can be calculated like this:
```

```go
func minimalBlocksize(messageLength uint64) uint64 {
    blocksize := NextHigherPowerOf2(messageLength)
    //Check if message fits with non-interactive default rules
    if messageLength <= (blocksize * (blocksize -1)) {
        return blocksize
    } else {
        return blocksize << 1
    }

> }

Reviewer 2: [question][proposal]

If we proceed with implementing this ADR, do we simultaneously want to increase https://github.com/celestiaorg/celestia-app/blob/76a275fd110ffb8990deca626a843780fd117733/pkg/appconsts/appconsts.go#L55 

and make this function default to a minimum square size of `appconsts.MinSquareSize`? @evan-forbes mentioned a higher minimum square size is desirable to keep message inclusion proofs small but we probably want to evaluate the table below to determine a reasonable minimum square size

minSquareSize | message size (bytes) | size of message inclusion proof (bytes)
--- | --- | ---
1 | |
2 | |
4 | |
8 | |
16 | |
32 | | 
64 | | 
128 | |
256 | |

Since a message inclusion proof needs to include all subtree roots for a message, if we don't increase the minimum square size, then I think small messages will result in large message inclusion proofs (to be confirmed).

Author: The table could have number of subroots instead of size of message inclusion proof as this is directly proportional right? I don`t think small messages will result in large message inclusion proofs

Reviewer 2: > and make this function default to a minimum square size of appconsts.MinSquareSize?

Disregard, I learned we don't want to do this and the proposed minimalBlocksize dependent on message size seems preferable.

> The table could have number of subroots instead of size of message inclusion proof as this is directly proportional right?

Yes, I think so.

```

## Status

> Proposed

Reviewer: imo, making a decision on the status of this should not be a blocker for merging. We should merge it regardless.

Author: @musalbas

## Consequences

### Negative

1. The amount of subtree roots per commitment is O(sqrt(n)), while n is the number of message shares. The worst case for the number of subtree roots is depicted in the diagram below - an entire block missing one share.  
  ![Interactive Commitment 2](./assets/complexity.png)  

> The worst case for the current implementation depends on the block size. If it is the worst block size, as in `minimalBlocksize`, it is O(sqrt(n)) as well. On the other hand, if the message is only in one row, then it is O(log(n)).

Reviewer 2: ```suggestion
The worst case for the current implementation depends on the block size. If it is the worst block size, as in `minimumSquareSize`, it is O(sqrt(n)) as well. On the other hand, if the message is only in one row, then it is O(log(n)).
```

> Therefore the height of the tree over the subtree roots is in this implementation O(log(sqrt(n))), while n is the number of message shares. In the current implementation, it varies from O(log(sqrt(n))) to O(log(log(n))) depending on the block size.

Reviewer 2: ```suggestion
Therefore the height of the tree over the subtree roots is in this implementation O(log(sqrt(n))), while n is the number of message shares. In the current implementation, it varies from O(log(sqrt(n))) to O(log(log(n))) depending on the square size.
```

> 2. With more subtree roots, the number of merkle proofs will increase. With deeper subtree roots, the size of the merkle proofs will increase. So instead of having a merkle proof from the `DataRoot` to the `originalSubRoot` you will need the merkle proof from `DataRoot` to `2^k` amount of `miniSubRoots` with `k` being the height difference of `originalSubRoot` and `miniSubRoots`. You can optimize this by having only **one** merkle proof from `DataRoot` to `originalSubRoot` and then calculate the `originalSubRoot` from the `k` `miniSubRoots`. Because the merkle proof is created after the block is published, we know the block size and, therefore, if this `originalSubRoot` exists or not.

Reviewer 2: ```suggestion
2. With more subtree roots, the number of merkle proofs will increase. With deeper subtree roots, the size of the merkle proofs will increase. So instead of having a merkle proof from the `DataRoot` to the `originalSubRoot` you will need the merkle proof from `DataRoot` to `2^k` amount of `miniSubRoots` with `k` being the height difference of `originalSubRoot` and `miniSubRoots`. You can optimize this by having only **one** merkle proof from `DataRoot` to `originalSubRoot` and then calculate the `originalSubRoot` from the `k` `miniSubRoots`. Because the merkle proof is created after the block is published, we know the square size and, therefore, if this `originalSubRoot` exists or not.
```

### Positive

> 1. A Rollup can include the commitment in the block header *before* posting to Celestia because it is size-independent and does not have to wait for Celestia to confirm the block size. In general, the rollup needs access to this commitment in some form to verify a message inclusion proof guaranteeing data availability, which Rollmint currently does not have access to.

Reviewer 2: ```suggestion
1. A Rollup can include the commitment in the block header *before* posting to Celestia because it is size-independent and does not have to wait for Celestia to confirm the square size. In general, the rollup needs access to this commitment in some form to verify a message inclusion proof guaranteeing data availability, which Rollmint currently does not have access to.
```

> 2. In turn, this would serve as an alternative to [ADR 007](https://github.com/celestiaorg/optimint/blob/main/docs/lazy-adr/adr-007-header-commit-to-shares.md)

Reviewer 2: [optional] to disambiguate with adr-007 in this repo

```suggestion
2. In turn, this would serve as an alternative to [rollmint/adr-007](https://github.com/celestiaorg/optimint/blob/main/docs/lazy-adr/adr-007-header-commit-to-shares.md)
```

3. Here is one scheme on how a Rollup might use this new commitment in the block header. Let's assume a Rollup that looks like this:  
  BH1 <-- BH2 <-- BH3 <-- BH4  
  The Messages that are submitted to Celestia could look like this:  
  Message 1: B1  
  The Commitment of B1 is saved into BH1.  
  Message 2: (BH1+B2)  
  The Commitment of BH1+B2 is saved into BH2.  
  Message 3: (BH2+B3)  
  The Commitment of (BH2+B3) is saved into BH3, and so on.  

> 4. Verifying a message inclusion proof could be done with merkle proofs of the subtree roots to the `DataRoot`, recalculating the commitment, and comparing to what's in the rollup block header. If you trust the validator set of Celestia/Rollup (aka Rollmint Light Client) it could be as simple as submitting a proof over the PFD transaction that included the message and then checking if the commitment is the same as in the PFD transaction.

Reviewer: I think you mentioned why this wasn't the case before but I forgot... Why can't this be done with the current approach? Don't we have to wait for it to be included in a block before we can create a proof for it?

> If you trust the validator set of Celestia/Rollup (aka Rollmint Light Client)

side note, but just to document that while this is currently the case, this will not be after state fraud proofs.

Author: Yes. You need to wait for the rollup message to be included in celestia to create a proof that it was included.
 For the verification you need to have committed to a commitment before it was posted to celestia. If you include all possible commitments in the rollup blockheader than verifying the commitment via the pfd transaction is possible as well. Currently we don't save any commitment to the header of the rollmint block.

Author: > side note, but just to document that while this is currently the case, this will not be after state fraud proofs.
wdym by that.

Reviewer: just that rollmint light clients shouldn't have to trust the validator set after we have fraud proofs for everything

5. So far, a full node in Rollmint downloads new blocks from the DA Layer after each Celestia block, coupled tightly for syncing. With this approach, we can send blocks over the p2p Layer giving a soft-commit to full nodes. Then, they would receive the hard-commit after verifying a message inclusion proof without the need to download the blocks anymore. **P2P Blocksync**

>     1. P2P Blocksync allows a Rollmint full node to run a Celestia light node and not a Celestia full node.

Author: That would mean that the blockheader would include as many commitments as possible blocksizes. Which is possible and still deterministic. Validation would take longer, because the subroots to verify the commitments would change, as you need to verify all subroot combinations. You are right you dont _need_ a single commitment.

## Context

Currently, commitments in Celestia are dependent on the size of the block. The following diagram visualizes how this can happen. The visualizations in this ADR are without parity shares/erasure encoding.
The yellow block in the diagram is a message of length 11 shares. A square of 4x4 results in the following subroots: `A B C D`.
In [`CreateCommit`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L112-166), we calculate the commitment of the subroots with [`HashFromByteSlices`](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/crypto/merkle/tree.go#L7-L21). This resulting merkleroot is `Com`.
In a square of 8x8, it results in the following subtree roots: `H1 C D`
[`HashFromByteSlices`](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/crypto/merkle/tree.go#L7-L21) results now in the merkleroot `Com'`, which is not equal to `Com`

![Old Commitment](./assets/size-dependent-commitment.png)

To have a commitment independent of the block size, you would need to create a merkle tree over subroots that are included in every possible block size.

> If we agree on producing the commitment in the `minimalBlocksize` the resulting commitment will be the same no matter the block size. *This is because all the subroots in the `minimalBlocksize` are also included in bigger block sizes*.

Author: Second Question: Close, as you can see in the implementation at the bottom it still needs to fit regarding the non-interactive-default. A message can only start in the second row. I mentioned that in #827 with additional visualizations.

> If we agree on producing the commitment in the `minimalBlocksize` the resulting commitment will be the same no matter the block size. *This is because all the subroots in the `minimalBlocksize` are also included in bigger block sizes*.

Author: First Question: Maybe "We agree on producing the commitment as if the message was put into the `minimalBlocksize` . Therefore the subtree roots will be the same and  the resulting commitment as well no matter the blocksize." explains it better.

You can see this in the updated version of the diagram above. Now the merkle root will stay the same. This property results from the block width and length being a power of 2.  

![New Commitment](./assets/size-independent-commitment.png)

> You can see in the Diagramm below that no matter how the message is split up into rows, the hashes of the subroots stay the same.

Reviewer 3: ```suggestion
You can see in the Diagram below that no matter how the message is split up into rows, the hashes of the subroots stay the same.
```

![Row Size Comparison](./assets/row-size-comparison.png)

This scheme works for interactive commitments as well as long as the index of when a message starts is the same in every block size.
For example, in the following diagram, you see the message starting at the second share and still having the same commitment. I marked the skipped blocks pink to show that both messages have the same starting index.

> ![Interactive Commitment](./assets/interactive-commitment.png)

Author: If we use the first 2 elements of the hash then we would create a subroot that does not connect to the `DataRoot`of Celestia. We cannot use or that else we would need the whole message to verify the commitment defeating the purpose.

> If the message starting point index is larger than the row of `minimalBlocksize`, you take the index mod `minimalBlocksize`. The `minimalBlocksize` is shown in green and the index in pink. In both cases, the commitment stays the same.

Author: The amount of commitments you have to calculate depends on the amount of messages in a block. This does not change in this proposal. The proposal just shows how to calculate the commitments it does not change the amount.

> If the message starting point index is larger than the row of `minimalBlocksize`, you take the index mod `minimalBlocksize`. The `minimalBlocksize` is shown in green and the index in pink. In both cases, the commitment stays the same.

Reviewer 3: got you, thanks a lot

![Interactive Commitment 2](./assets/interactive-commitment2.png)

## Alternative Approaches

TODO

## Decision

TODO

## Detailed Design

To implement this decision, you need to change [`CreateCommit`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L112-166).
In Detail, [`powerOf2MountainRange`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L142) should take `minimalBlocksize` as an argument instead of `squareSize`.

Following the non-interactive default rules, `minimalBlocksize` can be calculated like this:

```go
func minimalBlocksize(mLength uint64) uint64 {
    blocksize := NextHigherPowerOf2(mLength)
    //Check if message fits with non-interactive-deafult rules
    if mLength <= (blocksize * (blocksize -1)) {
        return blocksize
    } else {
        return blocksize << 1
    }
}
```

## Status

Proposed

## Consequences

### Negative

> 1. The amount of subtroots per commitment increases from O(log(n)) to O(sqrt(n)) while n is the number of shares of a message. The worst case for the most amount of shares is depicted in the diagram below - an entire block missing one share.  

Reviewer 2: [question] why can't the last share of the last row be occupied by the message? In other words, for a square size of 8 (same as diagram) why is the maximum message length `(8 * 7) - 1 = 55` instead of `(8 * 7) = 56`?

> 1. The amount of subtroots per commitment increases from O(log(n)) to O(sqrt(n)) while n is the number of shares of a message. The worst case for the most amount of shares is depicted in the diagram below - an entire block missing one share.  

Reviewer 2: Oh thanks for clarifying! 
```suggestion
1. The amount of subtree roots per commitment increases from O(log(n)) to O(sqrt(n)) while n is the number of shares of a message. The worst case for number of subtree roots is depicted in the diagram below - an entire block missing one share.  
```

>   ![Interactive Commitment 2](./assets/complexity.png)

Reviewer 2: [question] in this diagram, I think the y-axis `O(sqrt(n))` is the number of subtree roots. What is the x-axis `O(log(sqrt(n))`?

Author: The x-axis is also the number of subtree roots. Here we go after the mountainRange Merkle Tree. So in this case the complexity is logarithmic to the size of the row (sqrt(n)). In this example it would be a maximum of 3 subtree roots.

Reviewer 2: Sorry I don't follow. What does the Merkle mountain range look like for this example and why are there only 3 subtree roots in it?

Author: I added the diagram above as they are the same example. The 3 subtree roots are A B C in this case.

2. With more subroots the amount of merkle proofs will increase. With deeper subroots the size of the merkle proofs will increase. So instead of having a merkle proof from the `DataRoot` to the `originalSubRoot` you will need the merkle proof from `DataRoot` to `2^k` amount of `miniSubRoots` with `k` being the height difference of `originalSubRoot` and `miniSubRoots`. You can optimize this by having one merkle proof from `DataRoot` to `originalSubRoot` and then `k` merkle proofs from `originalSubRoot` to the `miniSubRoots`. Because the merkle proof is created after the block is published, we know the block size and, therefore, if this `originalSubRoot` exists or not, like in smaller block sizes.

### Positive

> 1. A Rollup can include the commitment in the block header *before* posting on Celestia because it is size-independent and does not have to wait for Celestia to confirm the block size. In general, the rollup needs access to this commitment in some form to verify a message inclusion proof guaranteeing data availability, which Rollmint currently does not have access to.

Reviewer 3: ```suggestion
1. A Rollup can include the commitment in the block header *before* posting to Celestia because it is size-independent and does not have to wait for Celestia to confirm the block size. In general, the rollup needs access to this commitment in some form to verify a message inclusion proof guaranteeing data availability, which Rollmint currently does not have access to.
```

2. In turn, this would serve as an alternative to [ADR 007](https://github.com/celestiaorg/optimint/blob/main/docs/lazy-adr/adr-007-header-commit-to-shares.md)
3. Here is one scheme on how a Rollup might use this new commitment in the block header. Let's assume a Rollup that looks like this:  
  BH1 <-- BH2 <-- BH3 <-- BH4  
  The Messages that are submitted to Celestia could look like this:  
  Message 1: B1  
  The Commitment of B1 is saved into BH1.  
  Message 2: (BH1+B2)  
  The Commitment of BH1+B2 is saved into BH2.  
  Message 3: (BH2+B3)  
  The Commitment of (BH2+B3) is saved into BH3, and so on.  

> 4. Verifying a message inclusion proof could be done with merkle proofs of the subroots to the `DataRoot` and then recalculating the commitment and comparing to what's in the rollup block header. If you trust the validator set of Celestia/Rollup (aka Rollmint Light Client) it could be as simple as submitting a proof over the pfd transaction that included the message and then check if the commitment is the same as in the pfd transaction.

Reviewer 3: ```suggestion
4. Verifying a message inclusion proof could be done with merkle proofs of the subroots to the `DataRoot` and then recalculating the commitment and comparing to what's in the rollup block header. If you trust the validator set of Celestia/Rollup (aka Rollmint Light Client) it could be as simple as submitting a proof over the PFD transaction that included the message and then check if the commitment is the same as in the PFD transaction.
```

5. So far, a full node in Rollmint downloads new blocks from the DA Layer after each Celestia block, coupled tightly for syncing. With this approach, we can send blocks over the p2p Layer giving a soft-commit to full nodes. Then, they would receive the hard-commit after verifying a message inclusion proof without the need to download the blocks anymore. **P2P Blocksync**
    1. P2P Blocksync allows a Rollmint full node to run a Celestia light node and not a Celestia full node.
    2. It allows the Rollup node to continue running after Celestia halts, relying on soft commits with no data availability.
    3. It gives the Rollup the option to run asynchronously to Celestia because you don't have to wait for new Celestia blocks/commitments of the messages.

> 6. Combining P2P Blocksync and the scheme in 3. we could have multiple rollup blocks in one Celestia block. It could look like this:  

Reviewer 3: ```suggestion
6. Combining P2P Blocksync and the scheme in 3, we could have multiple rollup blocks in one Celestia block. It could look like this:  
```
